### PR TITLE
Speed up Hypervisor finding procedure

### PR DESCRIPTION
**Pull request by Woellchen**
Dramatically speed up the process by doing an additional HV filtering
round based on real data before HVs are locked in the next round to
yield it (locked).
Reordering of HV checks in order to execute cheap checks first and
expensive ones last. This can potentially reduce a lot of wasted time.

### CODE REVIEW

File: igvm/hypervisor.py [hunk 1]
```diff
@@ -215,6 +190,40 @@ def check_vm(self, vm, offline):
                     .format(*hw_pair)
                 )
 
+        # HV in supported state?
+        if self.dataset_obj['state'] not in ['online', 'online_reserved']:
```
Comment (kofrezo) [inline igvm/hypervisor.py]: I think this decision should be done when filtering for the hypervisors to check in the first place. This would make it impossible to change the hypervisor selection from the commandline with arguments such as --ignore-reserved.

See: https://github.com/innogames/igvm/blob/master/igvm/commands.py#L312

File: igvm/hypervisor.py [hunk 2]
```diff
@@ -215,6 +190,40 @@ def check_vm(self, vm, offline):
...
+            raise InvalidStateError(
+                'Hypervisor "{}" is not in online state ({}).'
+                .format(self.fqdn, self.dataset_obj['state'])
+            )
+
+        # Enough CPUs?
+        if vm.dataset_obj['num_cpu'] > self.dataset_obj['num_cpu']:
```
Comment (kofrezo) [inline igvm/hypervisor.py]: I think we allow CPU over allocation no ?

File: igvm/hypervisor.py [hunk 3]
```diff
@@ -215,6 +190,40 @@ def check_vm(self, vm, offline):
...
+            raise HypervisorError(
+                'Not enough CPUs. Destination Hypervisor has {0}, '
+                'but VM requires {1}.'
+                .format(self.dataset_obj['num_cpu'], vm.dataset_obj['num_cpu'])
+            )
+
+        # Proper VLAN?
+        if not self.get_vlan_network(vm.dataset_obj['intern_ip']):
```
Comment (kofrezo) [inline igvm/hypervisor.py]: This check also has taken place already see: 

https://github.com/innogames/igvm/blob/fdc85aa8825a07abf24448df0b7081f817d91f80/igvm/commands.py#L833

File: igvm/hypervisor.py [hunk 4]
```diff
@@ -215,6 +190,40 @@ def check_vm(self, vm, offline):
...
+                'Hypervisor "{}" does not support route_network "{}".'
+                .format(self.fqdn, vm.route_network)
+            )
+
+        # Those checks below all require libvirt connection,
+        # so execute them last to avoid unnecessary overhead if possible.
+
+        # Enough memory?
+        free_mib = self.free_vm_memory()
+        if vm.dataset_obj['memory'] > free_mib:
```
Comment (kofrezo) [inline igvm/hypervisor.py]: This one is arguable. I think it is safe for us but maybe this should still remain a hypervisor preference because people might want to allow memory ballooning. What is your thought on this @TuxPowered42

File: igvm/commands.py [hunk 5]
```diff
@@ -4,14 +4,19 @@
...
+from collections import OrderedDict
 from contextlib import contextmanager, ExitStack
+from ipaddress import ip_address
+from os import environ
 
 from adminapi.dataset import Query
 from adminapi.filters import Any, StartsWith, Contains
 from fabric.colors import green, red, white, yellow
+from fabric.context_managers import settings
+from fabric.decorators import parallel
```
Comment (TuxPowered42) [inline igvm/commands.py]: Any chance of doing this via standard Python methods? I would prefer to keep Fabric dependencies at minimum.

File: igvm/commands.py [hunk 6]
```diff
@@ -78,10 +83,12 @@ def evacuate(hv_hostname, offline=None, dry_run=False):
...
-            if (
-                offline is not None and
-                (offline == [] or vm_function in offline)
-            ):
+            is_offline_migration = (
+                offline is not None
+                and (offline == [] or vm_function in offline)
+            )
+
+            if is_offline_migration:
```
Comment (TuxPowered42) [inline igvm/commands.py]: Please split cosmetic changes into a separate commit.

File: igvm/commands.py [hunk 7]
```diff
@@ -827,57 +838,60 @@ def _get_hypervisor(hostname, allow_reserved=False):
...
+    for possible_hv in possible_hvs.values():
         try:
-            hypervisor.check_vm(vm, offline)
-        except libvirtError as error:
-            hypervisor.release_lock()
-            log.warning(
-                'Preferred hypervisor "{}" is skipped: {}'.format(
-                    hypervisor, error)
-            )
+            possible_hv.acquire_lock()
```
Comment (kofrezo) [inline igvm/commands.py]: Will this lock be release if an exception is thrown ?

File: igvm/commands.py [hunk 8]
```diff
@@ -827,57 +838,60 @@ def _get_hypervisor(hostname, allow_reserved=False):
...
-            )
+
+        if not _check_vm(possible_hv, vm, offline):
+            possible_hv.release_lock()
             continue
 
         try:
-            yield hypervisor
+            yield possible_hv
+            return
```
Comment (kofrezo) [inline igvm/commands.py]: yield return makes no sense. You can just return.

File: igvm/commands.py (hunk 7 above)
Comment (Woellchen) [reply igvm/commands.py]: good catch. fill fix

Comment (Woellchen) [reply igvm/commands.py]: actually when this throws an exception there is in fact no lock. so nothing to unlock

File: igvm/commands.py (hunk 8 above)
Comment (Woellchen) [reply igvm/commands.py]: true story

File: igvm/hypervisor.py (hunk 1 above)
Comment (TuxPowered42) [reply igvm/hypervisor.py]: Interesting observation. But this code is then even more wrong: a HV could change its state whille it's being searched for. We should check here for exactly requested state.

File: igvm/hypervisor.py (hunk 2 above)
Comment (TuxPowered42) [reply igvm/hypervisor.py]: Noppe.

File: igvm/commands.py (hunk 5 above)
Comment (Woellchen) [reply igvm/commands.py]: yeah you can probably fiddle with `multiprocessing` or similar.. but no please not :D fabric makes it real easy in this case to parallelize

File: igvm/hypervisor.py (hunk 3 above)
Comment (TuxPowered42) [reply igvm/hypervisor.py]: Same as with state, we check a HV for all parameters while its locked. However Serveradmin users could change this without locking.

If we don't care for that, then this is redundant code, yes.

File: igvm/hypervisor.py (hunk 1 above)
Comment (Woellchen) [reply igvm/hypervisor.py]: lets wait for feedback from @TuxPowered42 but addresse this in another pr since the logic is the same as before, just in another place

File: igvm/hypervisor.py (hunk 2 above)
Comment (Woellchen) [reply igvm/hypervisor.py]: lets wait for feedback from @TuxPowered42 but addresse this in another pr since the logic is the same as before, just in another place

File: igvm/hypervisor.py (hunk 3 above)
Comment (Woellchen) [reply igvm/hypervisor.py]: lets wait for feedback from @TuxPowered42 but addresse this in another pr since the logic is the same as before, just in another place

File: igvm/hypervisor.py (hunk 4 above)
Comment (Woellchen) [reply igvm/hypervisor.py]: lets wait for feedback from @TuxPowered42 but addresse this in another pr since the logic is the same as before, just in another place

Comment (TuxPowered42) [reply igvm/hypervisor.py]: We don't allow for memory ballooning. We under-allocate VMs and give them their own swap.

File: igvm/commands.py (hunk 8 above)
Comment (TuxPowered42) [reply igvm/commands.py]: With the style change of "for ... else: throw" to "for ... return; throw" this seems necessary. Why the style change in the first place?

File: igvm/hypervisor.py (hunk 3 above)
Comment (TuxPowered42) [reply igvm/hypervisor.py]: If it's redundant we could make a separate PR to fix it. I would prefer not to change it too much with other changes @Woellchen did.

File: igvm/commands.py (hunk 8 above)
Comment (Woellchen) [reply igvm/commands.py]: actually i agree to @kofrezo. this is not a generator. it returns exactly one hypervisor. i will remove the yield statement

Comment (Woellchen) [reply igvm/commands.py]: i just reiterated on this. `yield` is necessary because this function is used as a context manager in order to make sure hv is always locked and unlocked correctly. i don't see a point to use an `else` statement. when the outer code is done working on the hv, this context manager is done, hence return

Comment (Woellchen) [reply igvm/commands.py]: i reverted to the `for/else` construct

File: igvm/commands.py (hunk 5 above)
Comment (Woellchen) [reply igvm/commands.py]: ok after discussion i have implemented a simple parallel executor based on threads without the use of fabric. this also prevents libvirt connections from being established more than once since fabric's `parallel` is based on subprocesses

**Comment by Woellchen** [comment]
ready when you are
